File: 조준장/골드5/2668.py
import sys


def dfs(start: int, num: int):
    visited.add(num)

    if arr[num] not in visited:
        dfs(start, arr[num])
    elif arr[num] in visited and start == arr[num]:
        answer.append(start)


n = int(sys.stdin.readline())
arr = [0] + [int(sys.stdin.readline()) for _ in range(n)]
answer = []
for i in range(1, n + 1):
    visited = set()
    dfs(i, i)

print(len(answer))
for a in answer:
    print(a)

File: 조준장/골드5/2668.kt
class `2668` {
    lateinit var arr: List<Int>
    lateinit var visited: BooleanArray
    fun solve() {
        val br = System.`in`.bufferedReader()
        val bw = System.out.bufferedWriter()
        val sb = StringBuilder()

        val n = br.readLine().toInt()

        arr = listOf(0) + List(n) { br.readLine().toInt() }
        val answer = mutableListOf<Int>()
        for (i in 1 .. n) {
            visited = BooleanArray(n + 1)
            if (dfs(i, arr[i])) {
                answer.add(i)
            }
        }

        sb.append(answer.size).append("\n")
        sb.append(answer.joinToString("\n"))
        bw.write(sb.toString())

        br.close()
        bw.close()
    }

    fun dfs(start: Int, num: Int): Boolean {
        if (start == num) {
            return true
        }

        if (!visited[num]) {
            visited[num] = true
            return dfs(start, arr[num])
        }

        return false
    }
}


fun main() {
    `2668`().solve()
}
